src/common/compat_strndup.c

```diff
@@ -0,0 +1,38 @@
+/*
+ * Copyright (c) 2023 Luka Logar <luka.logar@iname.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#ifndef HAVE_STRNDUP
+#include <stdlib.h>
+#include <string.h>
+#include "common/compat_strnlen.h"
+
+char *strndup(const char *str, size_t n)
+{
+	char *dst;
+	n = strnlen(str, n);
```

Comment:
please check `str != NULL`

Reply:
Since `str` is `const char *` I get an `error: ‘nonnull’ argument ‘str’ compared to NULL [-Werror=nonnull-compare]` error during compiling?

I could drop the const qualifier or probably get around this by using a hackish cast, but other string helper functions here also don't do any kind of input parameter checking?

Reply:
Could it be that you're using some other header file for strndup, which specifies `__attribute__((nonnull))` for `str`?

Reply:
Nope. 
It turned out, `gcc` has some (lots of, see [Other-Builtins](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html)) built-in functions with it's own declarations which in case of `strndup` specify ` __attribute__((nonnull))` for `str`. 
But oddly enough, at least in the case of MinGW `x86_64-w64-mingw32-gcc`, there is only a built-in `strndup` declaration and not also accompanying definition.
Using  `-fno-builtin` gets rid of this error, but using this looks sub-optimal to me? 
Or we can just leave it as it is, as I mentioned before, no other `compat_str*` function does any parameter checking (quite possibly due to the same `nonnull-compare` error)?

Reply:
normally, this should only be compiled in if ./configure doesn't detect `strndup`. If it is not detected while being built into gcc/mingw, then the detection mechanism is faulty.

Reply:
I think the `./configure` `strndup()` detection works just fine.
It detects `strndup()` on Linux & co. and doesn't detect `strndup()` on MinGW (as it should).

I think on both platforms `gcc` has built-in `strndup()` declaration but on Linux it then links to `strndup()` from `glibc` (or something like that) and since on Windows there is no system defined `strndup()` (or standard C library) the `strndup()` declaration was simply removed from `string.h` and `strndup()` not implemented (but the built-in `strndup()` declaration was kept in `gcc`)

In any case even the "standard" `glibc` [strndup()](https://github.com/bminor/glibc/blob/master/string/strndup.c) implementation doesn't check for `str != NULL`, so I suggest we leave it as it is...

Reply:
Not sure if we need to redo the mistakes of the past. I cannot find any documentation stating that this input must be nonnull.

Maybe you can check for strndup by linking against it instead of checking the definition in configure.ac

Reply:
A check for `strndup()` works just fine. And `strndup()`  definitely doesn't exist under MinGW-w64 but the compiler nonetheless throws an error as it has a leftover built-in `strndup()` declaration. At least that's how I see it.

So I just renamed `strndup()` to `_strndup()` and use that instead...